### PR TITLE
Add support for `--force-if-includes` to force push more safely

### LINKED ISSUE #144635: Add --force-if-includes option or default to make git push --force-with-lease setting safer
Add --force-if-includes setting or default to prevent unexpected work loss when --force-with-lease is enable in settings.

From official git docs:

> A general note on safety: supplying this option without an expected value, i.e. as --force-with-lease or --force-with-lease=<refname> interacts very badly with anything that implicitly runs git fetch on the remote to be pushed to in the background, e.g. git fetch origin on your repository in a cronjob.

Other git settings such as "Autofetch" and "Fetch on pull" make vscode especially likely to unexpectedly drop remote changes when performing a force-with-lease push.

Eloquent explanation of flags, tradeoffs, and why unexpected losses could occur: https://stackoverflow.com/a/65839129/1331851

### LINKED ISSUE #190356: Force push overwrites other people's commits when autofetch is enabled
Does this issue occur when all extensions are disabled?: Yes

Related: #144635

Version: 1.81.1 (user setup)
Commit: 6c3e3dba23e8fadc360aed75ce363ba185c49794
Date: 2023-08-09T22:22:42.175Z
Electron: 22.3.18
ElectronBuildId: 22689846
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Windows_NT x64 10.0.22621 (Windows 11)

Steps to Reproduce:

1. Rebase or commit amend
2. Wait for other people to push their commits
3. <kbd>Ctrl</kbd> + <kbd>,</kbd>
4. Turn “Git: Allow Force Push” & “Git: Autofetch” on
5. Wait for Git to fetch other people's commits
6. <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd>
7. “Git: Push (Force)”

“Git: Autofetch” is asked if it is enabled by a toast dialog.

https://ruky.me/2021/02/21/dont-allow-vscode-to-auto-fetch-from-repos/

This is because VS Code doesn't use `--force-if-includes` with `--force-with-lease`.

Self-service: a fix has been submitted in #187932.

### PR DESCRIPTION
**Pull request by tats-u**
Fixes #144635 
Fixes #190356

`--force-if-includes` makes the force push even safer if used with `--force-with-lease`.
This option is available since Git 2.30.0. (Released on Dec 2020. The version in Ubuntu 22.04 is newer.)

`man git-push` says:

> Force an update only if the tip of the remote-tracking ref has been integrated locally.
>
> This option enables a check that verifies if the tip of the remote-tracking ref is reachable from one of the "reflog" entries of the local branch based in it for a rewrite. The check ensures that any updates from the remote have been incorporated locally by rejecting the forced update if that is not the case.
>
> If the option is passed without specifying --force-with-lease, or specified along with --force-with-lease=<refname>:<expect>, it is a "no-op".

[The release note](https://github.com/git/git/blob/master/Documentation/RelNotes/2.30.0.txt) says:

> "git push --force-with-lease[=<ref>]" can easily be misused to lose commits unless the user takes good care of their own "git fetch".
> A new option "--force-if-includes" attempts to ensure that what is being force-pushed was created after examining the commit at the tip of the remote ref that is about to be force-replaced.

`git push --force-with-lease --force-if-includes` is so long that we hesitate to type.

VS Code asks us if it turns "Git: Autofetch" on by a toast:

https://ruky.me/2021/02/21/dont-allow-vscode-to-auto-fetch-from-repos/

Innocent Git newbies will press "Yes" and get ready for overwriting others' remote commits by innocent force push incompletely taught with rebase by Git "expert" bad at teaching.

This PR is created based on #53286.

image

image

### CODE REVIEW

**Comment by tats-u** [comment]
@lszomoru Could you assign another person if you're too busy to review my PR?

**Comment by tats-u** [comment]
@lszomoru Only `--force-with-lease` can't protect others' pushes when the auto-fetch is enabled.
**Why on earth do you ignore this PR?**

**Comment by tats-u** [comment]
@lszomoru A month is passed. If you don't want to review my change, bring someone else ***as soon as possible***.

**Comment by tats-u** [comment]
@joaomoreno @alexr00 @bpasero 
Dear maintainers who have reviewed recent PRs of `is:pr is:merged assignee:lszomoru -author:lszomoru Git`:
@lszomoru seems to be too busy to give me a reply for more than one month, so could you review this PR in favor of him?
He hasn't review any PRs for more than one month.

**Comment by tats-u** [comment]
@joaomoreno _When on earth_ can I get a co l review from one of you?

File: extensions/git/package.json
Comment (lszomoru) [inline extensions/git/package.json:2574]: Please switch to `markdownDescription`.

File: extensions/git/package.nls.json
Comment (lszomoru) [inline extensions/git/package.nls.json:220]: Please add the following at the end of the description to be consistent with other settings:
" **Note:** Requires git version `2.30.0` or later."

File: extensions/git/src/commands.ts
Comment (lszomoru) [inline extensions/git/src/commands.ts:3672]: Should the second sentence be "Check if force push is not prohibited by the remote or pull the latest changes from the remote branch first."

File: extensions/git/src/git.ts
Comment (lszomoru) [inline extensions/git/src/git.ts:1920]: We should only add this flag if the user is running a version of git that supports it. You can use `this.compareGitVersionTo('2.30')` similar to how it is used for other git commands.

**Review by lszomoru — CHANGES_REQUESTED** [review]
Could you please address my comments? Thank you!

File: extensions/git/package.nls.json
Comment (tats-u) [reply extensions/git/package.nls.json:220]: I'll make "git" capitalized ("Git"). How about this?

File: extensions/git/src/commands.ts
Comment (tats-u) [reply extensions/git/src/commands.ts:3672]: I'll fully comply with your suggestion.

File: extensions/git/package.json
Comment (tats-u) [reply extensions/git/package.json]: I only have to change "d" to "markdownD" in "description".

File: extensions/git/src/git.ts
Comment (tats-u) [reply extensions/git/src/git.ts:1920]: @lszomoru If user's Git is older, do you think it is desired to fall back to the behavior of only `--force-with-lease`?

File: extensions/git/package.nls.json
Comment (tats-u) [reply extensions/git/package.nls.json]: I have wanted to add after your note:

> If your Git is older, this option does not make sense.

File: extensions/git/src/git.ts
Comment (tats-u) [reply extensions/git/src/git.ts:1920]: I found the `_git` field of `Git` class, which has the `compareGitVersionTo` instance, in the class that has that method. I'll use `this._git.compareGitVersion` instead.

**Comment by tats-u** [comment]
@lszomoru I finished the modification. Could you review my changes again?
Also if you want me to squash commits into one, you can feel free to tell it.

**Review by lszomoru — APPROVED** [review]

**Comment by lszomoru** [comment]
@tats-u, I have pushed a commit to make some minor improvements.

**Review by aeschli — APPROVED** [review]

File: extensions/git/src/api/git.d.ts
Comment (tats-u) [inline extensions/git/src/api/git.d.ts:364]: @lszomoru Don't we have to prepare an error message for the rejection of the `--force` push as before?
Also I wonder why you split `SaferForcePushRejected` into `ForcePushLease(IfIncludes)Rejected`.
The only difference between them is how safe they are. What you have to do when they are rejected are same. Do you have a future plan to split error messages for them?

File: extensions/git/src/commands.ts
Comment (tats-u) [inline extensions/git/src/commands.ts:3665]: @lszomoru I believe the pull operation make the matter worse in the case the user intentionally choose the force push instead of the usual one.
How about changing "`Try running "Pull" first to pull the latest changes from the remote branch first.`" to "`Reflect the latest changes from the latest changes from branch first e.g. by trying running "Pull" first.`"?

File: extensions/git/src/commands.ts
Comment (tats-u) [reply extensions/git/src/commands.ts]: @lszomoru Could you tell me why did you tell me to change "`rebase to`" to "`pull`"?

Comment (tats-u) [reply extensions/git/src/commands.ts]: @lszomoru Is this just because Git itself makes such suggestion? Well, in that case, it's no wonder...

File: extensions/git/package.nls.json
Comment (lszomoru) [reply extensions/git/package.nls.json:220]: No need. I will hide these settings for users that do not have the correct Git version installed.

File: extensions/git/src/api/git.d.ts
Comment (lszomoru) [reply extensions/git/src/api/git.d.ts:364]: There is a concept of a push error handler where an extension can leverage the git extension API to register a push error handler. At the moment there are no push error handlers related to errors during force push, but in the future there might be so we want to make sure that the push error handler can distinguish between the two cases. Hope that makes sense.

File: extensions/git/src/commands.ts
Comment (lszomoru) [reply extensions/git/src/commands.ts:3665]: > I believe the pull operation make the matter worse in the case the user intentionally choose the force push instead of the usual one.

I think that the "theme" of these changes was to make force push safer and avoid data loss. In light of that, I think that the suggestion to Pull makes sense. Users that really want to force push will have to either change their settings or use the terminal. 

I am not sure I understand what you mean by "Reflect the latest changes from the latest change". I would suggest that we go ahead with the current message in the PR and then we can adjust it based on user feedback.

**Pull request merged**
